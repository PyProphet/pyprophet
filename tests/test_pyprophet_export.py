from __future__ import print_function

import os
import subprocess
import shutil
import sys

import pandas as pd
import sqlite3

import pytest

pd.options.display.expand_frame_repr = False
pd.options.display.precision = 4
pd.options.display.max_columns = None

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")

def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += str(subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT))
    except subprocess.CalledProcessError as error:
        print(error, end="", file=sys.stderr)
        raise
    return stdout


def _run_osw(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # peptide-level
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=global"

    # protein-level
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=global"

    # export
    cmdline += " export --in=test_data.osw --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    if not peptide:
        cmdline += " --no-peptide"

    if not protein:
        cmdline += " --no-protein"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_osw_unscored(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)

    # export
    cmdline = "pyprophet export --in=test_data.osw --out=test_data.tsv --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_ipf(regtest, temp_folder, transition_quantification=False, ipf="disable"):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms1 --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # MS2-level
    cmdline += " score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # transition-level
    cmdline += " score --in=test_data.osw --level=transition --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # IPF
    cmdline += " ipf --in=test_data.osw"

    # export
    cmdline += " export --in=test_data.osw --no-peptide --no-protein --ipf_max_peptidoform_pep=1  --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    cmdline += " --ipf=" + ipf

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_unscored(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # export
    cmdline = "pyprophet export-compound --in=test_data_compound.osw --out=test_data_compound_unscored.tsv --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_unscored.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_compound_ms1(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS1-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms1 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms1.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms1.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_ms2(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS2-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms2 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms2.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms2.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)


def test_osw_0(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, False)

def test_osw_1(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, True, False, False)

def test_osw_2(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, True, False)

def test_osw_3(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, True)

def test_ipf_0(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "disable")

def test_ipf_1(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, True, "disable")

def test_ipf_2(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "peptidoform")

def test_ipf_3(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "augmented")

def test_compound_0(tmpdir, regtest):
    _run_compound_ms1(regtest, tmpdir.strpath)
    
def test_compound_1(tmpdir, regtest):
    _run_compound_ms2(regtest, tmpdir.strpath)

def test_compound_unscored(tmpdir, regtest):
	_run_compound_unscored(regtest, tmpdir.strpath)
	
def test_osw_unscored(tmpdir, regtest):
	_run_osw_unscored(regtest, tmpdir.strpath)
