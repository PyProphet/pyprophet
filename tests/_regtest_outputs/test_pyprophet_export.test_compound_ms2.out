   Adducts  Charge   Intensity       RT  aggr_prec_Peak_Apex  aggr_prec_Peak_Area  assay_RT  assay_rt  compound_name  d_score  decoy  delta_RT  delta_rt                                           filename                   id  leftWidth  m_score        mz  peak_group_rank  rightWidth               run_id sum_formula  transition_group_id
0     M+H+       1   6225.5100  112.742              73086.0          314988.6293   117.983  117.9827  Methamidophos   2.9779      0    -5.241   -5.2406  /Users/alka/Documents/work/projects/OpenSWATH_...  6629051171722998253    108.687   0.0033  142.0086                1     121.853  3140099155998074833   C2H8NO2PS                    0
1     M+H+       1    758.0860  117.364              15110.0           49705.2486   117.983  117.9831  Methamidophos   2.9977      0    -0.619   -0.6191  /Users/alka/Documents/work/projects/OpenSWATH_...  8877729345763666378    114.448   0.0033  142.0086                1     121.854  2408145804652532658   C2H8NO2PS                    0
2     M+H+       1    198.6610  114.092               2967.0           12507.2948   117.983  117.9832  Methamidophos   1.9586      0    -3.891   -3.8912  /Users/alka/Documents/work/projects/OpenSWATH_...    85932679098247514    110.333   0.0219  142.0086                1     119.385  1007350642398073598   C2H8NO2PS                    0
3     M+H+       1    165.0120  117.559               2930.0            7454.3224   117.983  117.9831  Methamidophos   2.4718      0    -0.424   -0.4241  /Users/alka/Documents/work/projects/OpenSWATH_...  8707546894788545451    115.270   0.0090  142.0086                1     120.208   214379053049545951   C2H8NO2PS                    0
4     M+H+       1     31.2721  115.696                505.0            1301.6900   117.983  117.9831  Methamidophos   2.0051      0    -2.287   -2.2871  /Users/alka/Documents/work/projects/OpenSWATH_...  8419023093353902826    112.802   0.0208  142.0086                1     120.208  6344784885616328495   C2H8NO2PS                    0
..     ...     ...         ...      ...                  ...                  ...       ...       ...            ...      ...    ...       ...       ...                                                ...                  ...        ...      ...       ...              ...         ...                  ...         ...                  ...
95    M+H+       1    133.2040  456.633               4737.0           13662.9190   456.737  456.7369       Molinate   1.6555      0    -0.104   -0.1039  /Users/alka/Documents/work/projects/OpenSWATH_...  6094669438926273156    453.482   0.0357  188.1104                1     460.065  6344784885616328495    C9H17NOS                   25
96    M+H+       1  37115.3000  143.845              20587.0           65994.6621   146.160  146.1605       Acephate   2.8307      0    -2.315   -2.3155  /Users/alka/Documents/work/projects/OpenSWATH_...  1446287227778771923    140.780   0.0045  184.0192                1     149.832  3140099155998074833  C4H10NO3PS                   27
97    M+H+       1   8140.1400  145.778               6678.0           18505.5670   146.160  146.1603       Acephate   2.8453      0    -0.382   -0.3823  /Users/alka/Documents/work/projects/OpenSWATH_...   726467501366503692    143.249   0.0045  184.0192                1     149.010  2408145804652532658  C4H10NO3PS                   27
98    M+H+       1   1688.6600  144.121               1632.0            4300.0667   146.160  146.1601       Acephate   2.5860      0    -2.039   -2.0391  /Users/alka/Documents/work/projects/OpenSWATH_...   829341238503978341    141.603   0.0071  184.0192                1     147.363  1007350642398073598  C4H10NO3PS                   27
99    M+H+       1    907.0830  145.815                738.0            1741.9189   146.160  146.1597       Acephate   1.8558      0    -0.345   -0.3447  /Users/alka/Documents/work/projects/OpenSWATH_...  4292801597472146447    143.249   0.0243  184.0192                1     149.009   214379053049545951  C4H10NO3PS                   27

[100 rows x 23 columns]
